nhs/data/join_on_sa1.py

```diff
@@ -0,0 +1,219 @@
+
+import fiona
+import geopandas as gpd
+import matplotlib.pyplot as plt
+import numpy as np
+import polars as pl
+import yaml
+from fiona.drvsupport import supported_drivers
+from loguru import logger
+from pathos.multiprocessing import ProcessingPool as Pool
+from pyproj import CRS
+
+# Set the SHAPE_RESTORE_SHX option
+supported_drivers["ESRI Shapefile"] = "rw"
+
+
+def read_config(config_path: str) -> dict:
+    """Read YAML configuration file."""
+    with open(config_path, "r") as file:
+        config = yaml.safe_load(file)
+    logger.info(f"Configuration loaded: {config}")
+    return config
+
+
+def read_shapefile(path: str) -> int:
+    """Read shapefile and return total number of rows."""
+    total_rows = len(pl.read_parquet(path))
+    logger.info(f"Total number of rows in the shapefile: {total_rows}")
+    return total_rows
+
+
+def read_csv(path: str) -> pl.DataFrame:
+    """Read CSV file and return as Polars DataFrame."""
+    df = pl.read_csv(path)
+    logger.info(df.head())
+    return df
+
+
+def prepare_points(df: pl.DataFrame, crs: str) -> Any:
+    """Prepare points data for geospatial analysis."""
+    if "LONGITUDE" in df.columns and "LATITUDE" in df.columns:
+        df = df.rename({"LONGITUDE": "x", "LATITUDE": "y"})
+
+    gdf = gpd.GeoDataFrame(
+        df.to_pandas(),
+        geometry=gpd.points_from_xy(df["x"], df["y"]),
+        crs="EPSG:7844",  # Assuming input is in WGS84
+    )  # type: ignore
+    projected_crs = CRS.from_string(crs)
+    return gdf.to_crs(projected_crs)
+
+
+def process_chunk(map_data: Any, chunk: Any) -> pl.DataFrame:
+    """Process a chunk of point data, joining with area data."""
+    pnts_with_area = gpd.sjoin(chunk, map_data, how="left", predicate="within")
+    missing_points = pnts_with_area[pnts_with_area["SA2_NAME21"].isnull()]
+
+    if not missing_points.empty:
+        nearest_join = gpd.sjoin_nearest(
+            missing_points[["geometry"]], map_data, how="left"
+        )
+        pnts_with_area.loc[missing_points.index, "SA2_NAME21"] = nearest_join[
+            "SA2_NAME21"
+        ]
+        pnts_with_area.loc[missing_points.index, "SA1_CODE21"] = nearest_join[
+            "SA1_CODE21"
+        ]
+
+    result = pnts_with_area[["x", "y", "SA2_NAME21", "SA1_CODE21"]]
+    result = result.rename(columns={"SA2_NAME21": "area", "SA1_CODE21": "area_code"})
+    return pl.DataFrame(
+        result.to_dict()
+    )  # Convert to dict first to ensure compatibility with Polars
+
+
+def parallel_process(pnts_gdf: Any, map_data: Any, num_cores: int) -> pl.DataFrame:
+    """Parallel process point data."""
+    chunks = np.array_split(pnts_gdf, num_cores)
+    with Pool(num_cores) as pool:
+        all_results = pool.map(lambda chunk: process_chunk(map_data, chunk), chunks)
+    return pl.concat(all_results)
+
+
+def random_distribution(df: pl.DataFrame, num_iterations: int = 10) -> pl.DataFrame:
+    """Randomly shuffle lat-long pairs within their respective areas."""
```

Comment:
heavily expand on doc-string. It's not clear what column(s) `df` should have and what `num_iterations` is used for. Also explain the reason for needing this function.

Reply:
![image](https://github.com/user-attachments/assets/d688b740-920f-4c44-801b-e73c66517b21)

I think the data from the census mightn't be truly random so it's here just in case.